src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentNodeWriter.cs

```diff
@@ -210,10 +210,22 @@ public override void WriteComponentTypeInferenceMethod(CodeRenderingContext cont
                     if (child is ComponentAttributeIntermediateNode attribute)
                     {
                         var typeName = attribute.TypeName;
-                        if (attribute.BoundAttribute != null && !attribute.BoundAttribute.IsGenericTypedProperty())
+                        if (attribute.BoundAttribute != null)
                         {
-                            typeName = "global::" + typeName;
+                            if (attribute.BoundAttribute.IsEventCallbackProperty() && attribute.TryParseEventCallbackTypeArgument(out var argument))
+                            {
+                                // TypeInference is generated in a different namespace than the component type. Consequently any references to a type
+                                // must be globally qualified to prevent conflicts with the local namespace.
+                                // For EventCallback<SomeType>, we have to globally qualify both EventCallback and SomeType
+
+                                typeName = $"global::{ComponentsApi.EventCallback.FullTypeName}<global::{argument}>";
+                            }
+                            else if (!attribute.BoundAttribute.IsGenericTypedProperty())
```

Comment:
This will fix the problem, but only special casing `EventCallback`. 

What happens when you have an `EventCallback<T>` as the type? Or has that already been dealt with at this point?

Reply:
I had to do something like this in [taghelper binder](https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs#L415) but didn't need to use Roslyn. Maybe you can do something like that here if you don't want to special case it.

Reply:
> What happens when you have an EventCallback<T> as the type? Or has that already been dealt with at this point?

The previous check specifically works for `EventCallback<T>` since we expect it to both be an `EventCallback` and have a typed argument.

Are there other generic types we codegen? I couldn't tell looking at the code.

Reply:
This specific issue still exists for sub-generics right? Aka i imagine it wouldn't handle `EventCallback<List<Person>>`

I'm not sure utilizing `TryParseEventCallbackTypeArgument` fulfills this requirement 100%

Reply:
That's true. I wonder if an easier fix would be to not code-gen TypeInference in the random namespace.

Reply:
Right, I expected this to be a part of the generic component pass. We have a piece of infrastructure for this, as well as for knowing the difference between the type parameters of a component and other stuff. 

I think we have all of the pieces to solve this, I'm not a fan of the approach being used here.